File: cpp/238.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, 1);
        int pre = 1, post = 1;
        for (int i = 1, j = n-1; i < n; ++i) {
            res[i] *= pre;
            res[j] *= post;
            pre *= res[i];
            post *= res[j];
        } // 正着求
        return res;
    }
};

File: python/238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1] * n
        l, r = 1, 1
        for i in range(0, n):
            j = n - i - 1
            res[i] *= l
            res[j] *= r
            l *= nums[i]
            r *= nums[j]
        return res